api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,616 @@
+		log.Printf("[DEBUG] [EnvironmentScaler] Environment '%s' is already at desired scale of '%d'.", clusterName, desiredScale)
+	}
+
+	// choose which instances to terminate during our scale-down process
+	// instead of having asg randomly select instances
+	// e.g. if we scale from 5 to 3, we can terminate up to 2 unused instances
+	if desiredScale > minScale {
+		for i, max := 0, currentScale-desiredScale; i < max && i < len(unusedProviders); i++ {
+			instanceID := unusedProviders[i].ID
+
+			log.Printf("[DEBUG] [EnvironmentScaler] Terminating unused instance '%s' from environment '%s'.", instanceID, clusterName)
+
+			if err := e.terminateInstanceInAutoScalingGroup(instanceID); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*scaler.ResourceProvider, error) {
+	var environment *models.Environment
+
+	val, ok := e.environmentCache.Getf(clusterName)
+
+	switch ok {
+	case true:
+		environment = val.(*models.Environment)
+
+	case false:
+		env, err := e.EnvironmentProvider.Read(clusterName)
+		if err != nil {
+			return nil, err
+		}
+
+		e.environmentCache.Add(clusterName, env)
+		environment = env
+	}
+
+	instanceSpec, ok := instanceSpecifications()[environment.InstanceType]
+	if !ok {
+		return nil, fmt.Errorf("[EnvironmentScaler] Instance type '%s' is not valid!", environment.InstanceType)
+	}
+
+	return scaler.NewResourceProvider(nil, instanceSpec.CPU, instanceSpec.Memory, "<new instance>", nil, nil, nil), nil
+}
+
+func (e *EnvironmentScaler) calculateScaleDown(clusterName string, resourceProviders []*scaler.ResourceProvider) []*scaler.ResourceProvider {
+	var unusedProviders []*scaler.ResourceProvider
+
+	for _, provider := range resourceProviders {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	return unusedProviders
+}
+
+func (e *EnvironmentScaler) calculateScaleUp(clusterName string, resourceProviders []*scaler.ResourceProvider, resourceConsumers []scaler.ResourceConsumer) ([]*scaler.ResourceProvider, []error) {
+	providerDistributions := []ProviderDistribution{}
+
+	sorters := []func([]*scaler.ResourceProvider, []scaler.ResourceConsumer){
+		func(p []*scaler.ResourceProvider, c []scaler.ResourceConsumer) {
+			sortProvidersByCPU(p)
+			sortProvidersByUsage(p)
+			sortConsumersByCPU(c)
+		},
+		func(p []*scaler.ResourceProvider, c []scaler.ResourceConsumer) {
+			sortProvidersByMemory(p)
+			sortProvidersByUsage(p)
+			sortConsumersByMemory(c)
+		},
+	}
+
+	for _, sort := range sorters {
+
+		// Because *ResourceProvider.SubtractResourcesFor() actually modifies the underlying
+		// provider object, we need to make a fresh copy of the list of providers for each
+		// pass through a sort-and-allocate operation.
+		resourceProvidersCopy := make([]*scaler.ResourceProvider, len(resourceProviders))
+		for i, provider := range resourceProviders {
+			resourceProvidersCopy[i] = scaler.NewResourceProvider(
```

Comment:
this should be handled by a function on the resoure provider object, or a separate helper